### PR TITLE
Dev

### LINKED ISSUE #372: Cache update from multiple stages
The current implementation of Caches updates is not considering the requirement that stages different from the one where the cache is stored can be used as a source for the update. 

Solution: Modify the Dictionary used to specify the Caches to use a Tuple in the key where the tuple reflects the stage names that can be used to update the cache.

### LINKED ISSUE #461: Status Initial Condition update error in corner cases
There is a potential situation for systems that have Thermal Units modeled with UnitCommitment and also minimum stable power 0 where the initial conditions update consistency checks can error if the UC model didn't decommit the unit but resulted in P = 0.0.

### LINKED ISSUE #478: Non-spinning reserves
Add non-spinning reserves. 
The basic formulation would be similar to `RangeReserve` but only for non-committed generators.
It can be a fairly simplified formulation, with the following constraints: 
* Minimum and/or maximum reserve requirement.
* Only allow generators not committed to contribute to the reserve.
* Bound the generation + reserve by generator capacity.

As a first implementation, no time/ramp constraint would be necessary -- later on, we can investigate the importance of those.

### LINKED ISSUE #514: Improved Support for some Day-Ahead energy - services co-optimization
The current design of devices and services doesn't support a straight forward way to make energy - services co-optimization problem. Several steps are required to support this in a straight forward way. I am opening this issue to keep track of the progress to get this going. 

- [x] New Cost Function structure  x-ref: https://github.com/NREL-SIIP/PowerSystems.jl/pull/596 
- [x] Refactor cost functions implementation x-ref: https://github.com/NREL-SIIP/PowerSimulations.jl/pull/516
- [x] Implement a system constructor finalizer call. 

cc. @andrewrosemberg @raphaelsaavedra

### LINKED ISSUE #640: Parameter Naming in Services
Service requirements have the word requirement twice. 

```julia
Parameters
==========

P__max_active_power__PowerLoad
service_requirement__requirement__VariableReserve_ReserveUp
P__max_active_power__RenewableDispatch
P__max_active_power__HydroDispatch
service_requirement__requirement__VariableReserve_ReserveDown
```

### LINKED ISSUE #648: Simulation Initial Condition handling
When running a simulation with a model that has ramp constraints on the devices, it is possible that the ramp constraints are infeasible if the initial conditions in the system aren't good. A worst case is when the initial value in the System is outside of the bounds of P_max/P_min. 

This is causing many confusions about the models. We could handle these situations with smarter heuristics for the system initialization.

### LINKED ISSUE #660: Load results empty
```
load_results!(res_uc, 1; initial_time = ts_stamp)
```

Returns empty dictionaries.

### LINKED ISSUE #670: get_all_constraint_index error for OperationsProblem
Getting an error using get_all_constraint_index
```
typeof(UC)
OperationsProblem{MultiStartUnitCommitmentCC}
```
```
get_all_constraint_index(UC)
ERROR: MethodError: no method matching optimizer_index(::Int64)
Closest candidates are:
  optimizer_index(::JuMP.VariableRef) at C:\Users\kdoubled\.julia\packages\JuMP\MIPb6\src\JuMP.jl:1026
  optimizer_index(::JuMP.ConstraintRef{JuMP.Model,C,Shape} where Shape<:JuMP.AbstractShape where C) at C:\Users\kdoubled\.julia\packages\JuMP\MIPb6\src\JuMP.jl:1043
Stacktrace:
 [1] get_all_constraint_index(::OperationsProblem{MultiStartUnitCommitmentCC}) at C:\Users\kdoubled\.julia\packages\PowerSimulations\yXxfT\src\core\operations_problem.jl:951
 [2] top-level scope at REPL[118]:1
```

### LINKED ISSUE #701: Issue with MarketBidCost when linking problems in simulation.
I’m trying to define a simulation in the following way to represent the UC (to solve for commitments) and EDs (to solve for prices) of a DA operation. However, the simulation is not updating correctly the ED values according to the time series defined for the MarketBidCost.

The simulation sequence:
```
problems = SimulationProblems(;
        UC=OperationsProblem(
            template_uc,
            system_uc;
            optimizer=solver_uc,
            system_to_file=system_to_file,
            kwargs...,
        ),
        ED=OperationsProblem(
            template_ed,
            system_ed;
            optimizer=solver_ed,
            system_to_file=system_to_file,
            services_slack_variables=services_slack_variables,
            balance_slack_variables=balance_slack_variables,
            constraint_duals=constraint_duals,
            kwargs...,
        ),
    )
    feedforward_chronologies = Dict(("UC" => "ED") => Synchronize(; periods=24))
    feedforward = Dict(
        ("ED", :devices, :ThermalStandard) =>
            SemiContinuousFF(; binary_source_problem=ON, affected_variables=[ACTIVE_POWER]),
    )
    intervals = Dict("UC" => (Hour(24), Consecutive()), "ED" => (Hour(1), Consecutive()))
    uc_ed_sequence = SimulationSequence(;
        problems=problems,
        intervals=intervals,
        ini_cond_chronology=InterProblemChronology(),
        feedforward_chronologies=feedforward_chronologies,
        feedforward=feedforward,
    )
```
The `MarketBidCost` "variable cost" SingleTimeSeries added to both UC and ED for generator `bus4_virtual_supply`:
image

The `MarketBidCost` "variable cost" Deterministic forecast in UC:
```
julia> get_time_series_values(Deterministic, gen_uc, "variable_cost")
24-element Vector{Float64}:
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 1.0e7
 0.0
 
 
 julia> get_time_series(Deterministic, gen_uc, "variable_cost")
DeterministicSingleTimeSeries time_series (7):
   single_time_series: SingleTimeSeries time_series (168)
   initial_timestamp: 2020-01-01T00:00:00
   interval: 24 hours
   count: 7
   horizon: 24
```

The `MarketBidCost` "variable cost" Deterministic forecast in to ED:
```
julia> get_time_series_values(Deterministic, gen_ed, "variable_cost"; start_time=DateTime("2020-01-01")+Hour(23))
1-element Vector{Float64}:
 0.0


julia> get_time_series(Deterministic, gen_ed, "variable_cost"; start_time=DateTime("2020-01-01")+Hour(23))
DeterministicSingleTimeSeries time_series (145):
   single_time_series: SingleTimeSeries time_series (145)
   initial_timestamp: 2020-01-01T23:00:00
   interval: 1 hour
   count: 145
   horizon: 1
```

The ED problem for hour 24 is using the cost hour 1:
```
julia> print(sim.problems[:ED].internal.optimization_container.JuMPmodel)
Min 3000 P__ThermalStandard_{Solitude, 1} + 1500 P__ThermalStandard_{Park City, 1} + 1.0e9 P__ThermalStandard_{bus4_virtual_supply, 1} + 1400 P__ThermalStandard_{Alta, 1} + 1000 P__ThermalStandard_{Brighton, 1} + 4000 P__ThermalStandard_{Sundance, 1}
```

### LINKED ISSUE #704: Add cost break-down and other possible summary stats
**Describe the solution you'd like**
Add some functions to make quick access to some summary statistics about the solution easier. In addition to 
`read_optimizer_stats`, a cost breakdown by variable cost, shut-down cost, slack penalty cost, etc. would be great.  Total renewable curtailment might be another one, along with other common "totals."

### LINKED ISSUE #717: `StaticReserve` broken
It appears that the `StaticReserve` model is not working in a simulation:
```julia
julia> ed_problem_template = OperationsProblemTemplate()
set_device_model!(ed_problem_template, ThermalStandard, ThermalDispatch)
set_device_model!(ed_problem_template, PowerLoad, StaticPowerLoad)
set_service_model!(ed_problem_template, StaticReserve{ReserveUp}, RampReserve)
problem = OperationsProblem(ed_
Operations Problem Specification
============================================
Transmission: CopperPlatePowerModel
============================================
Devices Models: 

problem_template, sys;
                                 No Models Specified

============================================
Branches Models: 

         No Models Specified

============================================
Services Models:

         No Models Specified
 optimizer=solver,
                            con
============================================


julia> straint_duals=[:CopperPlateBalance,
julia> set_device_model!(ed_problem_template, ThermalStandard, ThermalDispatch)

julia> set_device_model!(ed_problem_template, PowerLoad, StaticPowerLoad)

julia> set_service_model!(ed_problem_template, StaticReserve{ReserveUp}, RampReserve)

julia> problem = OperationsProblem(ed_problem_template, sys;
                                   optimizer=solver,
                                   constraint_duals=[:CopperPlateBalance,
                                                                                 :requirement__VariableReserve_ReserveUp],
                                                                                                             horizon=1, balance_slack_variables=true
                                                                                                                                         )

Operations Problem Specification
============================================
Transmission: CopperPlatePowerModel
============================================
Devices Models: 

        Type: ThermalStandard
        Formulation: ThermalDispatch

        Type: PowerLoad
        Formulation: StaticPowerLoad

============================================
Branches Models: 

         No Models Specified

============================================
Services Models:

        Type: StaticReserve{ReserveUp}
        Formulation: RampReserve

============================================

julia> 

julia> sim_problem = SimulationProblems(ED=problem)
SimulationProblems(OrderedCollections.OrderedDict{Symbol, OperationsProblem}(:ED => OperationsProblem()
), [:ED])

julia> sim_sequence = SimulationSequence(
           problems=sim_problem,
           intervals=Dict("ED" => (Minute(5), Consecutive())),
           #ini_cond_chronology = IntraProblemChronology(),
       )
┌ Warning: No Feedforward Chronologies have been defined. This configuration assummes that there is no information passing between problems
└ @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation_sequence.jl:103
┌ Warning: Single problem detected, the default Initial Condition Chronology is IntraProblemChronology(), other values will be ignored.
└ @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation_sequence.jl:109
Feed Forward Chronology
-----------------------

Initial Condition Chronology
----------------------------

1



julia> sim_dir = mktempdir()
"/var/folders/27/2jr8c7gn4j72fvrg4qt81zrw8w_711/T/jl_UxeLkW"

julia> sim = Simulation(
           name="economic_dispatch",
           steps=1,
           problems=sim_problem,
           sequence=sim_sequence,
           simulation_folder=sim_dir
       )
Simulation()


julia> 

julia> build!(sim; serialize=true, console_level=Logging.Info)
[ Info: Building problem 1-ED
[ Info: Initializing Optimization Container
[ Info: Unit System changed to UnitSystem.SYSTEM_BASE = 0
┌ Warning: Data doesn't contain contributing devices with ramp limits for service SR, consider adjusting your formulation
└ @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/services_models/reserves.jl:218
┌ Error: Operation Problem Build Failed
│   exception =
│    The model container has invalid values
│    Stacktrace:
│      [1] error(s::String)
│        @ Base ./error.jl:33
│      [2] check_optimization_container(optimization_container::PowerSimulations.OptimizationContainer)
│        @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/optimization_container.jl:670
│      [3] build_impl!(optimization_container::PowerSimulations.OptimizationContainer, template::OperationsProblemTemplate, sys::System)
│        @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/optimization_container.jl:732
│      [4] problem_build!
│        @ ~/.julia/packages/PowerSimulations/65VAG/src/core/operations_problem.jl:520 [inlined]
│      [5] macro expansion
│        @ ~/.julia/packages/PowerSimulations/65VAG/src/core/operations_problem.jl:476 [inlined]
│      [6] macro expansion
│        @ ~/.julia/packages/TimerOutputs/ZQ0rt/src/TimerOutput.jl:220 [inlined]
│      [7] _build!(problem::OperationsProblem{GenericOpProblem}, serialize::Bool)
│        @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/operations_problem.jl:473
│      [8] _build_problems!(sim::Simulation, serialize::Bool)
│        @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:557
│      [9] macro expansion
│        @ ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:591 [inlined]
│     [10] macro expansion
│        @ ~/.julia/packages/TimerOutputs/ZQ0rt/src/TimerOutput.jl:220 [inlined]
│     [11] _build!(sim::Simulation, serialize::Bool)
│        @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:590
│     [12] (::PowerSimulations.var"#113#114"{Bool, Simulation})()
│        @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:647
│     [13] with_logstate(f::Function, logstate::Any)
│        @ Base.CoreLogging ./logging.jl:491
│     [14] with_logger
│        @ ./logging.jl:603 [inlined]
│     [15] macro expansion
│        @ ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:646 [inlined]
│     [16] macro expansion
│        @ ~/.julia/packages/TimerOutputs/ZQ0rt/src/TimerOutput.jl:220 [inlined]
│     [17] build!(sim::Simulation; output_dir::Nothing, recorders::Vector{Any}, console_level::Base.CoreLogging.LogLevel, file_level::Base.CoreLogging.LogLevel, serialize::Bool)
│        @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:626
│     [18] top-level scope
│        @ REPL[82]:1
│     [19] eval
│        @ ./boot.jl:360 [inlined]
│     [20] eval_user_input(ast::Any, backend::REPL.REPLBackend)
│        @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:139
│     [21] repl_backend_loop(backend::REPL.REPLBackend)
│        @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:200
│     [22] start_repl_backend(backend::REPL.REPLBackend, consumer::Any)
│        @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:185
│     [23] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool)
│        @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:317
│     [24] run_repl(repl::REPL.AbstractREPL, consumer::Any)
│        @ REPL /Users/julia/buildbot/worker/package_macos64/build/usr/share/julia/stdlib/v1.6/REPL/src/REPL.jl:305
│     [25] (::Base.var"#874#876"{Bool, Bool, Bool})(REPL::Module)
│        @ Base ./client.jl:387
│     [26] #invokelatest#2
│        @ ./essentials.jl:708 [inlined]
│     [27] invokelatest
│        @ ./essentials.jl:706 [inlined]
│     [28] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
│        @ Base ./client.jl:372
│     [29] exec_options(opts::Base.JLOptions)
│        @ Base ./client.jl:302
│     [30] _start()
│        @ Base ./client.jl:485
└ @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/operations_problem.jl:485
ERROR: Problem ED failed to build succesfully
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:33
  [2] _build_problems!(sim::Simulation, serialize::Bool)
    @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:559
  [3] macro expansion
    @ ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:591 [inlined]
  [4] macro expansion
    @ ~/.julia/packages/TimerOutputs/ZQ0rt/src/TimerOutput.jl:220 [inlined]
  [5] _build!(sim::Simulation, serialize::Bool)
    @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:590
  [6] (::PowerSimulations.var"#113#114"{Bool, Simulation})()
    @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:647
  [7] with_logstate(f::Function, logstate::Any)
    @ Base.CoreLogging ./logging.jl:491
  [8] with_logger
    @ ./logging.jl:603 [inlined]
  [9] macro expansion
    @ ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:646 [inlined]
 [10] macro expansion
    @ ~/.julia/packages/TimerOutputs/ZQ0rt/src/TimerOutput.jl:220 [inlined]
 [11] build!(sim::Simulation; output_dir::Nothing, recorders::Vector{Any}, console_level::Base.CoreLogging.LogLevel, file_level::Base.CoreLogging.LogLevel, serialize::Bool)
    @ PowerSimulations ~/.julia/packages/PowerSimulations/65VAG/src/core/simulation.jl:626
 [12] top-level scope
    @ REPL[82]:1

julia> 
```

### PR DESCRIPTION
**Pull request by claytonpbarrows**
This is to start making comments on things that don't quite work yet in v15

closes #372 
closes #461 
closes #478
closes #514 
closes #640 
closes #648 
closes #660
closes #670 
closes #717 
closes #704 
closes #701  
closes #802

### CODE REVIEW

File: src/operation/problem_template.jl [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+"""
+Sets the transmission model in a template.
+"""
+function set_transmission_model!(
+    template::ProblemTemplate,
+    model::NetworkModel{<:PM.AbstractPowerModel},
+)
+    template.transmission = model
+    return
+end
```
Comment (claytonpbarrows) [inline src/operation/problem_template.jl:64]:
```
"""
Sets the network model in a template.
"""
function set_network_model!(
    template::ProblemTemplate,
    model::NetworkModel{<:PM.AbstractPowerModel},
)
    template.network_model= model
    return
end
```

**Comment by claytonpbarrows** [comment]
this example fails: https://github.com/NREL-SIIP/SIIPExamples.jl/blob/master/script/3_PowerSimulations_examples/01_operations_problems.jl#L101 

```
julia> build!(op_problem, output_dir = mktempdir())
Academic license - for non-commercial use only - expires 2021-12-31
┌ Error: DecisionModel Build Failed
│   exception =
│    AssertionError: model.internal.ic_model_container === nothing
│    Stacktrace:
│      [1] build_initial_conditions!(model::DecisionModel{GenericOpProblem})
│        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/operation_model_interface.jl:110
│      [2] handle_initial_conditions!(model::DecisionModel{GenericOpProblem})
│        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/operation_model_interface.jl:174
│      [3] macro expansion
│        @ ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:247 [inlined]
│      [4] macro expansion
│        @ ~/.julia/packages/TimerOutputs/SSeq1/src/TimerOutput.jl:236 [inlined]
│      [5] build_pre_step!(model::DecisionModel{GenericOpProblem})
│        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:225
│      [6] build_impl!(model::DecisionModel{GenericOpProblem})
│        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/operation_model_interface.jl:211
│      [7] macro expansion
│        @ ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:278 [inlined]
│      [8] macro expansion
│        @ ~/.julia/packages/TimerOutputs/SSeq1/src/TimerOutput.jl:236 [inlined]
│      [9] (::PowerSimulations.var"#59#60"{DecisionModel{GenericOpProblem}})()
│        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:277
│     [10] with_logstate(f::Function, logstate::Any)
│        @ Base.CoreLogging ./logging.jl:511
│     [11] with_logger
│        @ ./logging.jl:623 [inlined]
│     [12] build!(model::DecisionModel{GenericOpProblem}; output_dir::String, recorders::Vector{Any}, console_level::Base.CoreLogging.LogLevel, file_level::Base.CoreLogging.LogLevel, disable_timer_outputs::Bool)
│        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:275
│     [13] top-level scope
│        @ REPL[79]:1
│     [14] eval
│        @ ./boot.jl:373 [inlined]
│     [15] eval_user_input(ast::Any, backend::REPL.REPLBackend)
│        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:150
│     [16] repl_backend_loop(backend::REPL.REPLBackend)
│        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:244
│     [17] start_repl_backend(backend::REPL.REPLBackend, consumer::Any)
│        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:229
│     [18] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool)
│        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:362
│     [19] run_repl(repl::REPL.AbstractREPL, consumer::Any)
│        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:349
│     [20] (::Base.var"#930#932"{Bool, Bool, Bool})(REPL::Module)
│        @ Base ./client.jl:394
│     [21] #invokelatest#2
│        @ ./essentials.jl:716 [inlined]
│     [22] invokelatest
│        @ ./essentials.jl:714 [inlined]
│     [23] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
│        @ Base ./client.jl:379
│     [24] exec_options(opts::Base.JLOptions)
│        @ Base ./client.jl:309
│     [25] _start()
│        @ Base ./client.jl:495
└ @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:285
BuildStatus.FAILED = 1
```

File: src/operation/problem_template.jl (hunk 1 above)
Comment (claytonpbarrows) [reply src/operation/problem_template.jl:64]: #795 fixes

**Comment by jd-lara** [comment]
> this example fails: https://github.com/NREL-SIIP/SIIPExamples.jl/blob/master/script/3_PowerSimulations_examples/01_operations_problems.jl#L101
> 
> ```
> julia> build!(op_problem, output_dir = mktempdir())
> Academic license - for non-commercial use only - expires 2021-12-31
> ┌ Error: DecisionModel Build Failed
> │   exception =
> │    AssertionError: model.internal.ic_model_container === nothing
> │    Stacktrace:
> │      [1] build_initial_conditions!(model::DecisionModel{GenericOpProblem})
> │        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/operation_model_interface.jl:110
> │      [2] handle_initial_conditions!(model::DecisionModel{GenericOpProblem})
> │        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/operation_model_interface.jl:174
> │      [3] macro expansion
> │        @ ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:247 [inlined]
> │      [4] macro expansion
> │        @ ~/.julia/packages/TimerOutputs/SSeq1/src/TimerOutput.jl:236 [inlined]
> │      [5] build_pre_step!(model::DecisionModel{GenericOpProblem})
> │        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:225
> │      [6] build_impl!(model::DecisionModel{GenericOpProblem})
> │        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/operation_model_interface.jl:211
> │      [7] macro expansion
> │        @ ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:278 [inlined]
> │      [8] macro expansion
> │        @ ~/.julia/packages/TimerOutputs/SSeq1/src/TimerOutput.jl:236 [inlined]
> │      [9] (::PowerSimulations.var"#59#60"{DecisionModel{GenericOpProblem}})()
> │        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:277
> │     [10] with_logstate(f::Function, logstate::Any)
> │        @ Base.CoreLogging ./logging.jl:511
> │     [11] with_logger
> │        @ ./logging.jl:623 [inlined]
> │     [12] build!(model::DecisionModel{GenericOpProblem}; output_dir::String, recorders::Vector{Any}, console_level::Base.CoreLogging.LogLevel, file_level::Base.CoreLogging.LogLevel, disable_timer_outputs::Bool)
> │        @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:275
> │     [13] top-level scope
> │        @ REPL[79]:1
> │     [14] eval
> │        @ ./boot.jl:373 [inlined]
> │     [15] eval_user_input(ast::Any, backend::REPL.REPLBackend)
> │        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:150
> │     [16] repl_backend_loop(backend::REPL.REPLBackend)
> │        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:244
> │     [17] start_repl_backend(backend::REPL.REPLBackend, consumer::Any)
> │        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:229
> │     [18] run_repl(repl::REPL.AbstractREPL, consumer::Any; backend_on_current_task::Bool)
> │        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:362
> │     [19] run_repl(repl::REPL.AbstractREPL, consumer::Any)
> │        @ REPL ~/Applications/Julia-1.7.app/Contents/Resources/julia/share/julia/stdlib/v1.7/REPL/src/REPL.jl:349
> │     [20] (::Base.var"#930#932"{Bool, Bool, Bool})(REPL::Module)
> │        @ Base ./client.jl:394
> │     [21] #invokelatest#2
> │        @ ./essentials.jl:716 [inlined]
> │     [22] invokelatest
> │        @ ./essentials.jl:714 [inlined]
> │     [23] run_main_repl(interactive::Bool, quiet::Bool, banner::Bool, history_file::Bool, color_set::Bool)
> │        @ Base ./client.jl:379
> │     [24] exec_options(opts::Base.JLOptions)
> │        @ Base ./client.jl:309
> │     [25] _start()
> │        @ Base ./client.jl:495
> └ @ PowerSimulations ~/Documents/repos/PowerSimulations.jl/src/operation/decision_model.jl:285
> BuildStatus.FAILED = 1
> ```

Fixed via 18757dd in the parameter_update branch.

File: src/simulation/simulation_models.jl [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+end
+
+function SimulationModels(;
+    decision_models,
+    emulation_model::Union{Nothing, EmulationModel} = nothing,
+)
+    return SimulationModels(decision_models, emulation_model)
+end
+
+function get_simulation_model(models::SimulationModels, name::Symbol)
```
Comment (claytonpbarrows) [inline src/simulation/simulation_models.jl]: can we make a method like this for `get_simulation_model(sim::Simulation, name::Symbol)`?

File: src/PowerSimulations.jl [hunk 3]
```diff
@@ -128,43 +129,50 @@ export run_economic_dispatch
...
-export get_existing_duals
-export get_existing_parameters
-export get_existing_timestamps
-export get_existing_variables
-export get_problem_name
+export get_dual_values
+export get_parameter_values
+export get_variable_values
+export get_timestamps
+export get_model_name
```
Comment (claytonpbarrows) [inline src/PowerSimulations.jl:142]: can we also export `get_jump_model`

File: src/simulation/simulation_models.jl (hunk 2 above)
Comment (claytonpbarrows) [reply src/simulation/simulation_models.jl]: ... and export this function

File: src/PowerSimulations.jl (hunk 3 above)
Comment (jd-lara) [reply src/PowerSimulations.jl:142]: The interface should be on `get_optimization_container`

File: src/simulation/simulation_models.jl (hunk 2 above)
Comment (jd-lara) [reply src/simulation/simulation_models.jl:41]: exported in ed40a47

File: src/devices_models/device_constructors/thermalgeneration_constructor.jl [hunk 4]
```diff
@@ -30,66 +46,114 @@ function construct_device!(
...
+        ReactivePowerBalance,
+        ReactivePowerVariable,
+        devices,
+        model,
+        S,
+    )
 
-    # Initial Conditions
-    initial_conditions!(optimization_container, devices, D())
+    add_expressions!(container, ProductionCostExpression, devices, model)
```
Comment (jd-lara) [inline src/devices_models/device_constructors/thermalgeneration_constructor.jl:77]: @claytonpbarrows we now populate this.

File: src/devices_models/devices/common/cost_functions.jl [hunk 5]
```diff
@@ -69,118 +76,132 @@ end
...
-    T_cf = typeof(optimization_container.cost_function)
+    T_cf = typeof(container.cost_function)
     if T_cf <: JuMP.GenericAffExpr && T_ce <: JuMP.GenericQuadExpr
-        optimization_container.cost_function += cost_expression
+        container.cost_function += cost_expression
     else
-        JuMP.add_to_expression!(optimization_container.cost_function, cost_expression)
+        JuMP.add_to_expression!(container.cost_function, cost_expression)
+    end
+    if has_expression(container, ProductionCostExpression, T)
```
Comment (jd-lara) [inline src/devices_models/devices/common/cost_functions.jl]: @claytonpbarrows here we add the cost expressions

Comment (claytonpbarrows) [reply src/devices_models/devices/common/cost_functions.jl]: @josephmckinsey , this is the new feature that will allow us to make an ex-poste calculation of generator costs. We'll need to make an extension for emissions to do the same thing for emissions costs

**Review by jd-lara — APPROVED** [review]
:tada:

**Pull request merged**